File: PyTorch/sparseconvnet/maxPooling.py
# Copyright 2016-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

from torch.autograd import Function, Variable
from torch.nn import Module
from .utils import *
from .sparseConvNetTensor import SparseConvNetTensor

class MaxPoolingFunction(Function):
    @staticmethod
    def forward(
            ctx,
            input_features,
            input_metadata,
            input_spatial_size,
            output_spatial_size,
            dimension,
            pool_size,
            pool_stride,
            nFeaturesToDrop):
        ctx.input_features=input_features
        ctx.input_metadata=input_metadata
        ctx.input_spatial_size = input_spatial_size
        ctx.output_spatial_size = output_spatial_size
        ctx.dimension = dimension
        ctx.pool_size = pool_size
        ctx.pool_stride = pool_stride
        ctx.nFeaturesToDrop = nFeaturesToDrop
        ctx.output_features = input_features.new()
        dim_typed_fn(dimension, input_features, 'MaxPooling_updateOutput')(
            input_spatial_size,
            output_spatial_size,
            pool_size,
            pool_stride,
            input_metadata.ffi,
            input_features,
            ctx.output_features,
            nFeaturesToDrop,
            torch.cuda.IntTensor() if input_features.is_cuda else nullptr)
        return ctx.output_features

    @staticmethod
    def backward(ctx, grad_output):
        grad_input=Variable(grad_output.data.new())
        dim_typed_fn(
            ctx.dimension, ctx.input_features, 'MaxPooling_updateGradInput')(
            ctx.input_spatial_size,
            ctx.output_spatial_size,
            ctx.pool_size,
            ctx.pool_stride,
            ctx.input_metadata.ffi,
            ctx.input_features,
            grad_input.data,
            ctx.output_features,
            grad_output.data,
            ctx.nFeaturesToDrop,
            torch.cuda.IntTensor() if ctx.input_features.is_cuda else nullptr)
        return grad_input, None, None, None, None, None, None, None


class MaxPooling(Module):
    def __init__(self, dimension, pool_size, pool_stride, nFeaturesToDrop=0):
        super(MaxPooling, self).__init__()
        self.dimension = dimension
        self.pool_size = toLongTensor(dimension, pool_size)
        self.pool_stride = toLongTensor(dimension, pool_stride)
        self.nFeaturesToDrop = nFeaturesToDrop
    def forward(self, input):
        output = SparseConvNetTensor()
        output.metadata = input.metadata
        output.spatial_size = (
            input.spatial_size - self.pool_size) / self.pool_stride + 1
        assert ((output.spatial_size-1)*self.pool_stride+self.pool_size==input.spatial_size).all()
        output.features =  MaxPoolingFunction().apply(
            input.features, input.metadata, input.spatial_size,
            output.spatial_size, self.dimension,self.pool_size,self.pool_stride,
            self.nFeaturesToDrop)
        return output
    def input_spatial_size(self, out_size):
        return (out_size - 1) * self.pool_stride + self.pool_size
    def __repr__(self):
        s = 'MaxPooling'
        if self.pool_size.max() == self.pool_size.min() and\
                self.pool_stride.max() == self.pool_stride.min():
            s = s + str(self.pool_size[0]) + '/' + str(self.pool_stride[0])
        else:
            s = s + '(' + str(self.pool_size[0])
            for i in self.pool_size[1:]:
                s = s + ',' + str(i)
            s = s + ')/(' + str(self.pool_stride[0])
            for i in self.pool_stride[1:]:
                s = s + ',' + str(i)
            s = s + ')'

        if self.nFeaturesToDrop > 0:
            s = s + ' nFeaturesToDrop = ' + self.nFeaturesToDrop
        return s

File: sparseconvnet/SCN/CUDA/MaxPooling.cpp
// Copyright 2016-present, Facebook, Inc.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

template <typename T>
void cuda_MaxPooling_ForwardPass(T *input_features, T *output_features,
                                 Int nPlanes, Int input_stride,
                                 Int output_stride, RuleBook _rules);
template <typename T>
void cuda_MaxPooling_BackwardPass(T *input_features, T *d_input_features,
                                  T *output_features, T *d_output_features,
                                  Int nPlanes, Int input_stride,
                                  Int output_stride, RuleBook _rules);

template <typename T, Int Dimension>
void cuda_MaxPooling_updateOutput(
    /*long*/ at::Tensor &inputSize, /*long*/ at::Tensor &outputSize,
    /*long*/ at::Tensor &poolSize,
    /*long*/ at::Tensor &poolStride, Metadata<Dimension> &m,
    /*cuda float*/ at::Tensor &input_features,
    /*cuda float*/ at::Tensor &output_features, long nFeaturesToDrop) {

  Int nPlanes = input_features.size(1) - nFeaturesToDrop;
  const auto &_rules =
      m.getRuleBook(inputSize, outputSize, poolSize, poolStride, true);
  Int nActive = m.getNActive(outputSize);
  output_features.resize_({nActive, nPlanes});
  output_features.zero_();

  auto iF = input_features.data_ptr<T>() + nFeaturesToDrop;
  auto oF = output_features.data_ptr<T>();
  cuda_MaxPooling_ForwardPass<T>(iF, oF, nPlanes, input_features.size(1),
                                 output_features.size(1), _rules);
}
template <typename T, Int Dimension>
void cuda_MaxPooling_updateGradInput(
    /*long*/ at::Tensor &inputSize, /*long*/ at::Tensor &outputSize,
    /*long*/ at::Tensor &poolSize,
    /*long*/ at::Tensor &poolStride, Metadata<Dimension> &m,
    /*cuda float*/ at::Tensor &input_features,
    /*cuda float*/ at::Tensor &d_input_features,
    /*cuda float*/ at::Tensor &output_features,
    /*cuda float*/ at::Tensor &d_output_features, long nFeaturesToDrop) {

  Int nPlanes = input_features.size(1) - nFeaturesToDrop;
  const auto &_rules =
      m.getRuleBook(inputSize, outputSize, poolSize, poolStride, true);
  d_input_features.resize_as_(input_features);
  d_input_features.zero_();

  auto iF = input_features.data_ptr<T>();
  auto oF = output_features.data_ptr<T>();
  auto diF = d_input_features.data_ptr<T>();
  auto doF = d_output_features.data_ptr<T>();
  cuda_MaxPooling_BackwardPass<T>(iF, diF, oF, doF, nPlanes,
                                  input_features.size(1),
                                  d_output_features.size(1), _rules);
}
template <typename T, Int Dimension>
void cuda_RandomizedStrideMaxPooling_updateOutput(
    /*long*/ at::Tensor &inputSize, /*long*/ at::Tensor &outputSize,
    /*long*/ at::Tensor &poolSize,
    /*long*/ at::Tensor &poolStride, Metadata<Dimension> &m,
    /*cuda float*/ at::Tensor &input_features,
    /*cuda float*/ at::Tensor &output_features, long nFeaturesToDrop) {

  Int nPlanes = input_features.size(1) - nFeaturesToDrop;
  const auto &_rules = m.getRandomizedStrideRuleBook(inputSize, outputSize, poolSize,
                                              poolStride, true);
  Int nActive = m.getNActive(outputSize);
  output_features.resize_({nActive, nPlanes});
  output_features.zero_();

  auto iF = input_features.data_ptr<T>() + nFeaturesToDrop;
  auto oF = output_features.data_ptr<T>();
  cuda_MaxPooling_ForwardPass<T>(iF, oF, nPlanes, input_features.size(1),
                                 output_features.size(1), _rules);
}
template <typename T, Int Dimension>
void cuda_RandomizedStrideMaxPooling_updateGradInput(
    /*long*/ at::Tensor &inputSize, /*long*/ at::Tensor &outputSize,
    /*long*/ at::Tensor &poolSize,
    /*long*/ at::Tensor &poolStride, Metadata<Dimension> &m,
    /*cuda float*/ at::Tensor &input_features,
    /*cuda float*/ at::Tensor &d_input_features,
    /*cuda float*/ at::Tensor &output_features,
    /*cuda float*/ at::Tensor &d_output_features, long nFeaturesToDrop) {

  Int nPlanes = input_features.size(1) - nFeaturesToDrop;
  const auto &_rules = m.getRandomizedStrideRuleBook(inputSize, outputSize, poolSize,
                                              poolStride, true);
  d_input_features.resize_as_(input_features);
  d_input_features.zero_();

  auto iF = input_features.data_ptr<T>();
  auto oF = output_features.data_ptr<T>();
  auto diF = d_input_features.data_ptr<T>();
  auto doF = d_output_features.data_ptr<T>();
  cuda_MaxPooling_BackwardPass<T>(iF, diF, oF, doF, nPlanes,
                                  input_features.size(1),
                                  d_output_features.size(1), _rules);
}
